nncf/telemetry/decorator.py

```diff
@@ -0,0 +1,74 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import inspect
+from typing import Callable
+from typing import List, Union
+
+from nncf.telemetry.events import get_current_category
+from nncf.telemetry.events import telemetry_category
+from nncf.telemetry.wrapper import NNCFTelemetry
+from nncf.telemetry.extractors import TelemetryExtractor
+from nncf.telemetry.extractors import VerbatimTelemetryExtractor
+
+
+class tracked_function:
+    """
+    Decorator to add telemetry events to a given function execution. The call to a decorated function
+    will in general result in a telemetry session being created and a set of events being sent before
+    function execution. The category of the session and events will be determined by parameters to the decorator.
+    """
+    def __init__(self, category: str = None, collectors: List[Union[str, TelemetryExtractor]] = None):
+        """
+        :param category: A category to be attributed to the events. If set to None, no events will be sent.
+        :param collectors: Add argument names in this list as string values to send an event with an "action" equal to
+            the argument name and "label" equal to the argument value before the function start.
+            The argument must be either a string or a dictionary of strings. If that is not the case, instead of
+            argument names you can specify an object of a customized `TelemetryExtractor` subclass; use the same
+            approach for more complex event reporting.
+        """
+        self._category = category
+        self._collectors = [VerbatimTelemetryExtractor(x) if isinstance(x, str) else x for x in collectors]
+
+    def __call__(self, fn: Callable) -> Callable:
+        fn_signature = inspect.signature(fn)
+
+        def wrapped(*args, **kwargs):
+            bound_args = fn_signature.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+            events = []  # type: CollectedEvent
+            for collector in self._collectors:
+                argname = collector.argname
+                argvalue = bound_args.arguments[argname] if argname is not None else None
+                event = collector.extract(argvalue)
+                events.append(event)
+
+            previous_category = get_current_category()
+            with telemetry_category(self._category) as category:
+                if category is not None:
+                    if category != previous_category:
+                        NNCFTelemetry.start_session(self._category)
+                    for event in events:
+                        NNCFTelemetry.send_event(event_category=category,
+                                                 event_action=event.name,
+                                                 event_label=event.data,
+                                                 event_value=event.int_data)
```

Comment:
Do we really need to send several events? I mean we send one event per method's argument in this approach. I thought we should send only one event per method call.

Reply:
While it seems to be possible to send a dict of something instead of separate key:value events, I'm not sure that it will be easy to analyze in the telemetry collection base. We'll see if it has a serious enough impact so that it needs to be optimized.

Reply:
I got it. Seems like some investigation is needed here to understand the usability of such approach.